File: src/main/java/isop/command_line/FindMethodAmongLexedTokens.java
package isop.command_line;

import isop.command_line.lex.Token;
import isop.command_line.lex.TokenType;
import isop.domain.Method;

import java.util.Collection;
import java.util.stream.Stream;

/**
 * Created by mathieu.
 */
public class FindMethodAmongLexedTokens {
    private static Stream<Method> getPotential(Collection<Method> methods, String methodName){
        return methods
                .stream()
                .filter(method -> method.name.equalsIgnoreCase(methodName))
                ;
    }

    public static Method findMethod(Collection<Method> methods, String methodName, Collection<Token> lexed)
    {
        long count = lexed.stream().filter(t -> t.tokenType.equals(TokenType.PARAMETER)).count();
        Method potentialMethod =
                getPotential(methods, methodName)
                .filter(method -> method.getParameters().size() <= count)
                .sorted((o1, o2) -> Integer.compare(
                        o1.getParameters().size(),
                        o2.getParameters().size()))
                .findAny()
                .orElse(null);
        if (potentialMethod != null)
        {
            return potentialMethod;
        }
        return getPotential(methods, methodName)
                .findAny()
                .orElse(null);
    }
}

File: src/main/java/isop/command_line/FindMethodAmongLexedTokens.kt
package isop.command_line

import isop.command_line.lex.Token
import isop.command_line.lex.TokenType
import isop.domain.Method
import java.util.stream.Stream

/**
 * Created by mathieu.
 */
object FindMethodAmongLexedTokens {
    private fun getPotential(methods: Collection<Method>, methodName: String): Stream<Method> {
        return methods
                .stream()
                .filter { method -> method.name.equals(methodName, ignoreCase = true) }
    }

    fun findMethod(methods: Collection<Method>, methodName: String, lexed: Collection<Token>): Method {
        val count = lexed.stream().filter { t -> t.tokenType == TokenType.PARAMETER }.count()
        val potentialMethod = getPotential(methods, methodName)
                .filter { method -> method.parameters.size <= count }
                .sorted { o1, o2 ->
                    Integer.compare(
                            o1.parameters.size,
                            o2.parameters.size)
                }
                .findAny()
                .orElse(null)
        return potentialMethod ?: getPotential(methods, methodName)
                .findAny()
                .orElse(null)
    }
}
